x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/ApiKeyService.java

```diff
@@ -192,14 +199,42 @@ public ApiKeyService(Settings settings, Clock clock, Client client, XPackLicense
         this.threadPool = threadPool;
         this.cacheHasher = Hasher.resolve(CACHE_HASH_ALGO_SETTING.get(settings));
         final TimeValue ttl = CACHE_TTL_SETTING.get(settings);
+        final Integer maximumWeight = CACHE_MAX_KEYS_SETTING.get(settings);
         if (ttl.getNanos() > 0) {
             this.apiKeyAuthCache = CacheBuilder.<String, ListenableFuture<CachedApiKeyHashResult>>builder()
                 .setExpireAfterWrite(ttl)
```

Comment:
We already have "expire after xxx" for all caches. The maximum weight is to control the overall size of a cache so that it does not grow unbounded. They are for different purposes.

Reply:
I think I haven't expressed myself clearly, what I meant:
```suggestion
                .setExpireAfterAccess(ttl)
```
I think the "expire after write" is suitable for the doc cache but the credential cache might benefit more from the "expire after access" strategy because the underlying data (the key secret) never changes.

Reply:
Ah OK I see your point now. This is existing code. I thought there was a reason why "expire after write" is used, but now I cannot remember any. I agree that "expire after access" seems to be a better fit. But I seem to remember noticing this previously, but I am not sure now why it was kept like this. I could be missing something. If you don't mind. I'll leave this out as future work.